### PR TITLE
Type::GetByName(): cache results for faster lookup

### PR DESCRIPTION
**Pull request by Al2Klimov**
Especially since #9543

## "Blocked by" (not really, but preferably)

* #9545

### CODE REVIEW

**Comment by julianbrost** [comment]
I just had a really quick look at this and find it strange doing it this way around.

How much work would it be to make the primary source of types a `std::unordered_map<String, Type::Ptr>` and make `globals.Types` a view into this?

**Comment by Al2Klimov** [comment]
Too much for what this change shall improve.

**Comment by Al2Klimov** [comment]
Also globals.System.Types is initialised very early and never changes.

**Comment by Al2Klimov** [comment]
## Before/after, `time icinga2 daemon -C`

f59f361f0
real    36m38.277s
user    196m31.058s
sys     71m49.837s

real    35m14.356s
user    190m16.353s
sys     68m23.075s

real    35m35.668s
user    192m36.516s
sys     68m44.489s

979a918a2
real    33m37.158s
user    185m54.210s
sys     61m20.227s

real    32m31.482s
user    182m6.965s
sys     58m28.920s

real    32m7.106s
user    176m20.685s
sys     60m13.250s

In short, I'm proud of this bridge technology.

**Comment by Al2Klimov** [comment]
Would you prefer if the register macros also fill a second, non-DSL, map (which is preferably accessed unlocked)?

**Comment by Al2Klimov** [comment]
> In short, I'm proud of this bridge technology.

We have caches also e.g. in #9408 after all.

**Comment by Al2Klimov** [comment]
* 56 cores/threads (IDK)
* 256G RAM
* Config of our customer 22470
* time icinga2 daemon -C

## master before #9627

real    54m20,817s
user    337m24,969s
sys     2459m21,103s

## master before #9627 + #9553

real    49m3,936s
user    345m6,072s
sys     2208m49,892s

## master with #9627

real    23m15,077s
user    1039m6,028s
sys     198m59,482s

## master with #9627 + #9553

real    21m52,338s
user    963m19,549s
sys     203m45,879s
